lgc/elfLinker/FetchShader.h

```diff
@@ -0,0 +1,87 @@
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  FetchShader.h
+ * @brief LGC header file: The class to generate the fetch shader used when linking a pipeline.
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "GlueShader.h"
+#include "lgc/state/PalMetadata.h"
+#include "lgc/state/PipelineState.h"
+
+namespace lgc {
+
+class LgcContext;
+
+// =====================================================================================================================
+// A fetch shader
+class FetchShader : public GlueShader {
+public:
+  FetchShader(PipelineState *pipelineState, llvm::ArrayRef<VertexFetchInfo> fetches,
+              const VsEntryRegInfo &vsEntryRegInfo);
+  ~FetchShader() override {}
+
+  // Get the string for this glue shader. This is some encoding or hash of the inputs to the create*Shader function
+  // that the front-end client can use as a cache key to avoid compiling the same glue shader more than once.
+  llvm::StringRef getString() override;
+
+  // Get the symbol name of the main shader that this glue shader is prolog or epilog for.
+  llvm::StringRef getMainShaderName() override;
+
+  // Get the symbol name of the glue shader.
+  llvm::StringRef getGlueShaderName() override {
+    return getEntryPointName(m_vsEntryRegInfo.callingConv, /*isFetchlessVs=*/false);
+  }
+
+  // Get whether this glue shader is a prolog (rather than epilog) for its main shader.
+  bool isProlog() override { return true; }
+
+  // Get the name of this glue shader.
+  llvm::StringRef getName() const override { return "fetch shader"; }
+
+  // No PAL metadata entries need updating for the fetch shader.
+  void updatePalMetadata(PalMetadata &) override { return; }
+
+protected:
+  // Generate the glue shader to IR module
+  llvm::Module *generate() override;
+
+private:
+  llvm::Function *createFetchFunc();
+
+  // The information stored here is all that is needed to generate the fetch shader. We deliberately do not
+  // have access to PipelineState, so we can hash the information here and let the front-end use it as the
+  // key for a cache of glue shaders.
+  llvm::SmallVector<VertexFetchInfo, 8> m_fetches;
+  VsEntryRegInfo m_vsEntryRegInfo;
```

Comment:
Can we initialize `VsEntryRegInfo` here?

Reply:
Won't it already be default initialized?  Is there something in the style guide that explains why it should be explicitly initialized?

Reply:
Let's leave this as is and do any non-obvious cleanup separately.